source/mqtt_subscription_set.c

```diff
@@ -0,0 +1,431 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "aws/mqtt/private/mqtt_subscription_set.h"
+
+#include "aws/mqtt/private/client_impl_shared.h"
+
+#define SUBSCRIPTION_SET_DEFAULT_BRANCH_FACTOR 10
+#define SUBSCRIPTION_SET_DEFAULT_ENTRY_COUNT 50
+
+struct aws_mqtt_subscription_set_subscription_record *aws_mqtt_subscription_set_subscription_record_new(
+    struct aws_allocator *allocator,
+    const struct aws_mqtt_subscription_set_subscription_options *subscription) {
+    struct aws_mqtt_subscription_set_subscription_record *record =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt_subscription_set_subscription_record));
+
+    record->allocator = allocator;
+    aws_byte_buf_init_copy_from_cursor(&record->topic_filter, allocator, subscription->topic_filter);
+    record->subscription_view = *subscription;
+    record->subscription_view.topic_filter = aws_byte_cursor_from_buf(&record->topic_filter);
+
+    return record;
+}
+
+void aws_mqtt_subscription_set_subscription_record_destroy(
+    struct aws_mqtt_subscription_set_subscription_record *record) {
+    if (record == NULL) {
+        return;
+    }
+
+    aws_byte_buf_clean_up(&record->topic_filter);
+    aws_mem_release(record->allocator, record);
+}
+
+static void s_aws_mqtt_subscription_set_subscription_record_hash_destroy(void *element) {
+    struct aws_mqtt_subscription_set_subscription_record *record = element;
+
+    aws_mqtt_subscription_set_subscription_record_destroy(record);
+}
+
+static struct aws_mqtt_subscription_set_topic_tree_node *s_aws_mqtt_subscription_set_node_new(
+    struct aws_allocator *allocator,
+    struct aws_mqtt_subscription_set_topic_tree_node *parent) {
+
+    struct aws_mqtt_subscription_set_topic_tree_node *node =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt_subscription_set_topic_tree_node));
+    node->allocator = allocator;
+    aws_hash_table_init(
+        &node->children,
+        allocator,
+        SUBSCRIPTION_SET_DEFAULT_BRANCH_FACTOR,
+        aws_hash_byte_cursor_ptr,
+        aws_mqtt_byte_cursor_hash_equality,
+        NULL,
```

Comment:
Maybe, but it would require a refactor that doesn't seem particularly useful.  The tree destroy rips it out of the parent's table anyways.  Hash table destructors can be useful in simple scenarios but we've had multiple fatal crash bugs due to their misuse and once you opt in to them it significantly constraints object lifetime management.